lib/experimental/lexopt.nim

```diff
@@ -0,0 +1,251 @@
+      LongOpt ## A long option
+      LongValue ## Value of a long option
+      ShortValue ## Value of a short option
+      Ended ## Nothing left
+
+  let currentState =
+    if l.index >= l.cmdline.len:
+      LexState.Ended
+    elif l.current.isLongOpt:
+      if l.valueIdx > 0:
+        LexState.LongValue
+      elif not l.current.isDashDash:
+        LexState.LongOpt
+      else:
+        LexState.Value
+    elif l.current.isShortOpt:
+      if l.valueIdx > 0 and l.current[l.valueIdx] in ValueDelims:
+        LexState.ShortValue
+      elif not l.current.isDash:
+        LexState.ShortOpt
+      else:
+        LexState.Value
+    else:
+      LexState.Value
+
+  case currentState
+  of LongOpt:
+    let delimIdx = l.current.find(ValueDelims)
+    if delimIdx > 2:
+      l.valueIdx = delimIdx
+      result = (cmdLong, l.current[2..<delimIdx])
+    else:
+      result = (cmdLong, l.current[2..^1])
+      inc l.index
+  of ShortOpt:
+    l.valueIdx = if l.valueIdx == 0: 1 else: l.valueIdx
+    result = (cmdShort, $l.current[l.valueIdx])
+
+    inc l.valueIdx
+    if l.valueIdx >= l.current.len:
+      l.valueIdx = 0
+      inc l.index
+  of Value:
+    result = (cmdValue, l.current)
+    inc l.index
+  of LongValue:
+    raise newUnexpectedValueError(l.current[0..<l.valueIdx], l.current[l.valueIdx + 1..^1])
+  of ShortValue:
+    raise newUnexpectedValueError("-" & $l.current[l.valueIdx - 1], l.current[l.valueIdx + 1..^1])
+  of Ended:
+    result = (cmdEnd, "")
+
+proc value*(l: var CmdLexer, delimitedOnly = false): Option[string] =
+  ## Consume the value for the current option, if any.
+  ##
+  ## The following section describes how various syntaxes are handled:
+  ##
+  ## 1. `--foo:bar` or `--foo=bar` yields `bar` if called after `next()`
+  ##    returns `foo`.
+  ## 2. `-abco:foo` or `-abco=foo` yields foo if called after `next()` returns
+  ##    `o`.
+  ## 3. If `delimitedOnly` is true, syntaxes below will not be considered.
+  ## 4. `-abcooutput.txt` yields `output.txt` if called after `next()` returns
+  ##    `o`.
+  ## 5. The current parameter is treated as a value, regardless of whether it
+  ##    can be considered as an option (ie. `--foo --bar` will yield `--bar` if
+  ##    this procedure is called after `--foo`).
+  if l.index < l.cmdline.len:
+    # The current index is parsed and contains a value
+    if l.valueIdx > 0:
+      # valueIdx is ValueDelims for either `--foo:bar` or `-o:bar`
+      if l.current[l.valueIdx] in ValueDelims:
+        result = some(l.current[l.valueIdx + 1..^1])
+
+      # This is a short option being parsed, return the remainder
+      # as a value if it is valid
+      elif not delimitedOnly:
+        result = some(l.current[l.valueIdx..^1])
+
+      # Do nothing otherwise
+      else:
+        return
+
+      l.valueIdx = 0
+      inc l.index
+
+    # The current index is not parsed, treat it as a value if valid
+    elif not delimitedOnly:
+      result = some(l.current)
+      inc l.index
```

Comment:
```suggestion
      elif not delimitedOnly:
        # This is a short option being parsed, return the remainder
        # as a value if it is valid
        result = some(l.current[l.valueIdx..^1])
      else:
        # Do nothing otherwise
        return

      l.valueIdx = 0
      inc l.index
    elif not delimitedOnly:
      # The current index is not parsed, treat it as a value if valid
      result = some(l.current)
      inc l.index
```
A purely stylistic suggestion. The spacing between the branches and the comments on the same indentation level are not common in this codebase.